File: src/main/scala/scala_examples/3_type_system.scala
package scala_examples

object TypeSystem extends App {
  val listOfStrings = List("1", "9", "11", "100")

  // Parse strings to integers
  val listOfIntegers = listOfStrings.map(i => i.toInt)

  // Filtering the list of integers works fine
  listOfIntegers.filter(i => i < 10).foreach(println)

  // Filtering the list of strings does not compile
  // TODO: uncomment; below code raises a compilation error
  // listOfStrings.filter(i => i < 10).foreach(println)
}

File: src/main/java/java_examples/3_type_system.java
package java_examples;

import java.util.*;

class TypeSystem {

    public static void main(String[] args) {
        List<String> listOfStrings = Arrays.asList("1", "9", "11", "100");

        // Parse strings to integers
        ArrayList<Integer> listOfIntegers = new ArrayList<>();
        for (String item : listOfStrings) {
            listOfIntegers.add(Integer.valueOf(item));
        }

        // Filtering the list of integers works fine
        for (Integer item: listOfIntegers) {
            if (item < 10) {
                System.out.println(item);
            }
        }

        // Filtering the list of strings does not compile
        // TODO: uncomment; below code raises a compilation error
//        for (String item: listOfStrings) {
//            if (item < 10) {
//                System.out.println(item);
//            }
//        }
    }
}
